archive/untar_stream_test.ts

```diff
@@ -0,0 +1,277 @@
+  assertEquals(entries.length, 0);
+});
+
+Deno.test("untarLinuxGeneratedTar", async function () {
+  const filePath = resolve(testdataDir, "deno.tar");
+  const file = await Deno.open(filePath, { read: true });
+
+  const expectedEntries = [
+    {
+      fileName: "archive/",
+      fileSize: 0,
+      fileMode: 509,
+      mtime: 1591800767,
+      uid: 1001,
+      gid: 1001,
+      owner: "deno",
+      group: "deno",
+      type: "directory",
+    },
+    {
+      fileName: "archive/deno/",
+      fileSize: 0,
+      fileMode: 509,
+      mtime: 1591799635,
+      uid: 1001,
+      gid: 1001,
+      owner: "deno",
+      group: "deno",
+      type: "directory",
+    },
+    {
+      fileName: "archive/deno/land/",
+      fileSize: 0,
+      fileMode: 509,
+      mtime: 1591799660,
+      uid: 1001,
+      gid: 1001,
+      owner: "deno",
+      group: "deno",
+      type: "directory",
+    },
+    {
+      fileName: "archive/deno/land/land.txt",
+      fileMode: 436,
+      fileSize: 5,
+      mtime: 1591799660,
+      uid: 1001,
+      gid: 1001,
+      owner: "deno",
+      group: "deno",
+      type: "file",
+      content: new TextEncoder().encode("land\n"),
+    },
+    {
+      fileName: "archive/file.txt",
+      fileMode: 436,
+      fileSize: 5,
+      mtime: 1591799626,
+      uid: 1001,
+      gid: 1001,
+      owner: "deno",
+      group: "deno",
+      type: "file",
+      content: new TextEncoder().encode("file\n"),
+    },
+    {
+      fileName: "archive/deno.txt",
+      fileMode: 436,
+      fileSize: 5,
+      mtime: 1591799642,
+      uid: 1001,
+      gid: 1001,
+      owner: "deno",
+      group: "deno",
+      type: "file",
+      content: new TextEncoder().encode("deno\n"),
+    },
+  ];
+
+  const untar = new UntarStream();
+  await file.readable.pipeTo(untar.writable);
+
+  for await (const entry of untar.readable) {
+    const expected = expectedEntries.shift();
+    assert(expected);
+    const content = expected.content;
+    delete expected.content;
+
+    // @ts-ignore its fine
+    assertEquals({ ...entry }, expected);
+
+    if (content) {
+      const buffer = new Buffer();
+      await entry.readable.pipeTo(buffer.writable);
+      assertEquals(content, buffer.bytes());
+    }
+  }
+});
+
+Deno.test("untarAsyncIteratorWithoutReadingBody", async function (): Promise<
+  void
+> {
+  const file = await Deno.open(filePath, { read: true });
+  const entries: TarOptions[] = [
+    {
+      name: "output.txt",
+      readable: ReadableStream.from([
+        new TextEncoder().encode("hello tar world!"),
+      ]),
+      contentSize: 16,
+    },
+    {
+      name: "dir/tar.ts",
+      readable: file.readable,
+      contentSize: file.statSync().size,
+    },
+  ];
+
+  const tar = ReadableStream.from<TarOptions>(entries).pipeThrough(new TarStream());
+
+  const untar = new UntarStream();
+  // read data from a tar archive
+  await tar.pipeTo(untar.writable);
+
+  for await (const entry of untar.readable) {
```

Comment:
when replacing this with a pipeThrough call instead, it breaks. needs to be figured out.